Added mysql_set_charset support

arraintxo opened the pull request:
PDO does not support setting the charset once the connection has been set, but we can use "SET NAMES xxx" as a workaround

arraintxo committed:
Added mysql_set_charset support

mattbit commented:
Hey, thanks! Seems good, I will check better and merge later.

mattbit commented:
Alright, after checking more thoroughly I found a problem: `mysql_set_charset` affects the encoding used for escaping with `mysql_real_escape_string`. Bypassing it by executing the query directly means that the escape function (`PDO::quote` in this case) is not aware of the correct charset. This may lead to subtle SQL injections as [described here](https://stackoverflow.com/questions/5741187/sql-injection-that-gets-around-mysql-real-escape-string/12118602#12118602).

Since this may have serious security implications, I would avoid introducing discrepancies with respect to the original behaviour.

Here is some code:
```php
// With the old mysql_set_charset
$link = mysql_connect('127.0.0.1', 'test_user');
var_dump(mysql_real_escape_string("\xbf\x27")); // string(3) "ø\'"

mysql_set_charset('gbk');
var_dump(mysql_real_escape_string("\xbf\x27")); // string(4) "\ø\'" (follows the new charset)

// Using PDO and SET NAMES query (note that PDO::quote also adds the quotes of course) 
$pdo = new PDO("mysql:host=127.0.0.1", "test_user");
var_dump($pdo->quote("\xbf\x27")); // string(5) "'ø\''"

$pdo->exec('SET NAMES gbk');
var_dump($pdo->quote("\xbf\x27")); // string(5) "'ø\''" (no change in the escaping)
```

The only solution I found so far is to integrate the charset in the DSN when building the PDO object:
```php
$pdo = new PDO("mysql:host=127.0.0.1;charset=gbk", "test_user");
var_dump($pdo->quote("\xbf\x27")); // string(6) "'\ø\''" (yes!)
```
In principle, when `mysql_set_charset` is used, we could replace the old PDO connection with a new one specifying the requested charset in the DSN. But this seems a bit too heavy in my opinion.

I would instead make the connection manager accessible, so that one would only need to replace the calls to `mysql_connect(…); mysql_set_charset(…);` with a PDO connection with the correct charset (e.g. `Mysql::getManager()->connect(…)` while keeping the other `mysql_*` functions in the rest of the code.

If you don’t have other ideas about how to fix this, I would close this pull request.
(Thanks a lot for the contribution in any case!)

arraintxo committed:
Mysql::getManager now allows retrieving the manager to stablish a custom DSN

arraintxo commented:
Understood.

Anyway we are obviously migrating from a pretty old system and need this functionallity while we migrate all the mysql calls, so I keep trying to have this properly working.

I updated the code to retrieve the manager through Mysql::getManager() method. I also added a simple test that checks if charset has been properly set.

Thanks for your work and I hope you like the current solution :)

mattbit commented:
Oh, this is nice!

To summarize, to use different charsets one needs to replace the bit of legacy code responsible for creating the connection and instead set up a shiny PDO connection directly. The rest of the code will continue to work as usual.

For example, the legacy code may look like
```php
mysql_connect('my_host', 'my_user', 'my_password');
mysql_set_charset('latin1');

// …

$result = mysql_query("SELECT * FROM users WHERE name = 'Friedrich Nietzsche'");
```
and must be replaced by manually setting up the PDO connection as follows:
```php
require __DIR__ . '/vendor/autoload.php';

use Mattbit\MysqlCompat\Mysql;
Mysql::defineGlobals(); // optional

// Set up the connection with the required PDO DSN (including the charset).
$manager = Mysql::getManager();
$manager->connect('mysql:dbhost=my_host;charset=latin1', 'my_user', 'my_password');

// …

// The rest works as usual, using the PDO connection created above.
$result = mysql_query("SELECT * FROM users WHERE name = 'Friedrich Nietzsche'");
